### PR TITLE
Optimize database sync process & optionally skip database analysis

### LINKED ISSUE #1586: Sync settings
Running Metabase on a big database creates an unnecessary load when it tries to sync the schema/data.
Having some sort of settings where one could disable or schedule the sync would help a lot.

### LINKED ISSUE #1701: Reduce number of background queries.
Hi team,

I was investigating an issue with my Redshift cluster consistently being at 100% CPU on the leader node. I found out 100% of queries consistently running in the background, such as:

```
SELECT AVG(CHAR_LENGTH(CAST("mk_3344_1"."events_raw_stage"."a_channelcounreferring_domain" AS CHAR))) AS "len" FROM "mk_3344_1"."events_raw_stage" 
```

```
SELECT COUNT(*) AS "count" FROM "mk_3337_1"."contacts" WHERE ("mk_3337_1"."contacts"."email" LIKE $1)
```

For every of the 100s of tables in my schema, all coming from my Metabase instance with no one using it.

I assume those are background queries to refresh the metadata. Is there a way to disable those or reduce their number / frequency? 
Redshift has a max concurrency, so having 1000s of queries queued in the background makes the cluster unusable.

Thanks!

### PR DESCRIPTION
**Pull request by agilliland**
fixes #1701 , fixes #1586 

Optimizations:
- Updated some parts of the IDriver interface to provide more generic table level functions which provide more room for each driver to optimize the introspection and analysis work.
- Improved the sync related functions in the generic_sql base driver to ensure the schema introspection process runs without requiring any row level data.  This allows even a larger DWH such as Redshift to sync just its schema in less than 30s.
- Reworked the way the mongo driver does its introspection so that we can scan the rows of data in a collection and extract most metadata in a single pass, including nested fields.  With this update syncing a mongo database is roughly 4-5x faster.
- Removed the automatic updating of FieldValues when a Field is inserted/updated.  We know only do this work as part of our table analysis functionality, which can optionally be skipped if desired.

Skipping Analysis:
- The sync process now accepts an optional argument `:full-sync?` which can be set to `false` if the caller wants to skip any parts of the sync process which are resource intensive and potentially slow.  This allows for much better sync performance at the expense of some of the metadata.

Other:
- Fixed a couple issues that were found with FK syncing.
- Updated table row counts to use a Long int so that we can handle big row counts.

### CODE REVIEW

**Comment by agilliland** [comment]
@camsaul Here's the code which updates the database sync process to give us some performance improvements on databases like Redshift and also provides a mechanism for users to skip the most costly parts of the sync process if they wanted.

@mazameli @kdoh This code is functionally ready to go at this point, so the main thing I need to finish this off will be a UI treatment for the toggle which lets the user opt out of the detailed sync work.

File: src/metabase/driver.clj [hunk 1]
```diff
@@ -273,7 +290,8 @@
         java.util.UUID               :TextField
         org.postgresql.util.PGobject :UnknownField} klass)
       (cond
-        (isa? klass clojure.lang.IPersistentMap) :DictionaryField)
+        (isa? klass clojure.lang.IPersistentMap)    :DictionaryField
+        (isa? klass clojure.lang.IPersistentVector) :ArrayField)
```
Comment (camsaul) [inline src/metabase/driver.clj]: Could use `condp` here instead
```clojure
(condp isa? klass
  IPersistentMap ...)
```

File: src/metabase/driver/redshift.clj [hunk 2]
```diff
@@ -24,27 +24,33 @@
...
+                                     JOIN pg_attribute source_column ON c.conrelid     = source_column.attrelid
+                                     JOIN pg_class dest_table        ON c.confrelid    = dest_table.oid
+                                     JOIN pg_namespace dest_table_ns ON dest_table.relnamespace = dest_table_ns.oid
+                                     JOIN pg_attribute dest_column   ON c.confrelid    = dest_column.attrelid
+                              WHERE c.contype            = 'f'::char
+                                     AND source_table.relname = ?
+                                     AND n.nspname            = ?
+                                     AND source_column.attnum = ANY(c.conkey)
+                                     AND dest_column.attnum   = ANY(c.confkey)"
+                             (:name table)
```
Comment (camsaul) [inline src/metabase/driver/redshift.clj]: this SQL is already so nicely indented can we keep the `AS` and `=` bits aligned?

File: src/metabase/models/field.clj [hunk 3]
```diff
@@ -36,7 +36,8 @@
 
 (def ^:const base-types
   "Possible values for `Field.base_type`."
-  #{:BigIntegerField
+  #{:ArrayField
+    :BigIntegerField
```
Comment (camsaul) [inline src/metabase/models/field.clj]: Do we support arrays now?

**Comment by camsaul** [comment]
:+1:

File: src/metabase/driver/redshift.clj (hunk 2 above)
Comment (agilliland) [inline src/metabase/driver/redshift.clj]: :+1:

**Pull request merged**